File: backend/lib/convex.py
import os
from typing import Any

import httpx


class ConvexUnavailable(Exception):
    """Raised when the server cannot safely reach the configured Convex deployment."""


async def call_convex(function_type: str, path: str, args: dict[str, Any]) -> Any:
    site_url = os.environ.get("CONVEX_SITE_URL") or os.environ.get("CONVEX_URL")
    token = os.environ.get("CONVEX_APP_TOKEN")
    if not site_url or not token:
        raise ConvexUnavailable("Convex is not configured")

    endpoint = f"{site_url.rstrip('/')}/api/{function_type}"
    payload = {"path": path, "args": {"token": token, **args}}
    try:
        async with httpx.AsyncClient(timeout=12.0) as client:
            response = await client.post(endpoint, json=payload)
            response.raise_for_status()
            body = response.json()
            if "value" not in body:
                raise ConvexUnavailable("Convex returned an invalid response")
            return body["value"]
    except (httpx.HTTPError, ValueError, KeyError) as exc:
        raise ConvexUnavailable("Convex is temporarily unavailable") from exc

File: dist-server/backend/services/convex.js
export class ConvexUnavailable extends Error {
    constructor(message = "Convex is unavailable") {
        super(message);
        this.name = "ConvexUnavailable";
    }
}
export async function callConvex(functionType, path, args) {
    const siteUrl = process.env.CONVEX_SITE_URL ?? process.env.CONVEX_URL;
    const token = process.env.CONVEX_APP_TOKEN;
    if (!siteUrl || !token)
        throw new ConvexUnavailable("Convex is not configured");
    try {
        const response = await fetch(`${siteUrl.replace(/\/$/, "")}/api/${functionType}`, {
            method: "POST",
            headers: { "content-type": "application/json" },
            body: JSON.stringify({ path, args: { token, ...args } }),
            signal: AbortSignal.timeout(12_000),
        });
        if (!response.ok)
            throw new ConvexUnavailable(`Convex returned HTTP ${response.status}`);
        const body = (await response.json());
        if (body.value === undefined)
            throw new ConvexUnavailable("Convex returned an invalid response");
        return body.value;
    }
    catch (error) {
        if (error instanceof ConvexUnavailable)
            throw error;
        throw new ConvexUnavailable("Convex is temporarily unavailable");
    }
}
